test/cocotb/osdtestlib/modtestclasses.py

```diff
@@ -0,0 +1,637 @@
+        y = len(message)
+        
+        if y > 8:
+            # output an error if too many flits are to be sent
+            raise TestFailure("Number of data words exceeds upper limit!")
+        
+        for x in range(0, y):
+            # extract one word after the other
+            newmessage = message[x]
+            
+            # determine if the last flit has to be sent
+            if x == (y-1):
+                # indicate this by setting the 'last'-bit
+                yield packet_driver.build_flit(dut = dut, data = newmessage, 
+                                               last = 1)
+            else:
+                yield packet_driver.build_flit(dut = dut, data = newmessage, 
+                                               last = 0)
+        dut.debug_in.valid <= 0
+        
+    @cocotb.coroutine
+    def receive_packet(self, dut, response):
+        """successively fills a buffer with incoming data words until the last 
+           flit of a packed is received
+        Args:
+            dut:            device under test.
+            response:       (preferably) empty buffer which the incoming flits
+                            are stored in.
+        Returns:
+            response of the debug module of variable length in the 
+            'response'-buffer
+        """
+        
+        # get response
+        while True:
+            # wait for the active clock edge
+            yield RisingEdge(dut.clk)
+            
+            if dut.debug_out.valid.value:
+                
+                # extend the buffer with the debug module's output
+                response.append(dut.debug_out.data.value.integer)
+                
+                # if the last flit has been sent
+                if dut.debug_out.last.value:
+                    # stop reading in the response 
+                    break
+                
+    @cocotb.coroutine        
+    def RegRead(self, dut, read_driver, read_utility, destination, source, 
+                reg_width, address, response):
+        """function which uses the subroutine 'send_packet' to read a value from
+           a specified register and reads in the response to tell the user if 
+           the read process was successful
+            
+        Args:
+            dut:            device under test.
+            destination:    id of the module which is the destination.
+            source:         id of the module which is the source.
+            reg_width:      choose between 16, 32, 64 and 128 bit register 
+                            access.
+            address:        address of the register the value is to be read 
+                            from.
+            response:       buffer variable which the response of the dut will 
+                            be stored in
+            
+        Returns:    
+            -
+        
+        """
+        
+        # branch depending on the width of the accessed data word
```

Comment:
The if branches differ only in the subtype and the expected response type. Set these two as variables depending on the width, and put all other code in the common code path (outside the ifs).